pkg/kube/dialer.go

```diff
@@ -0,0 +1,442 @@
+	// of the default namespace)
+	switch len(labels) {
+	case 1:
+		if !isServiceDNS {
+			pod.Namespace = defaultNamespace
+		} else {
+			return pod, 0, ErrInvalidPodFQDN
+		}
+	case 2:
+		if isServiceDNS {
+			pod.Namespace = defaultNamespace
+		} else {
+			pod.Namespace = labels[1]
+		}
+	case 3:
+		pod.Namespace = labels[2]
+	default:
+		return pod, 0, ErrInvalidPodFQDN
+	}
+
+	pod.Name = labels[0]
+
+	return pod, port, nil
+}
+
+func (p *PodDialer) connectionForPodLocked(pod types.NamespacedName) (*refCountedConnection, error) {
+	if conn, ok := p.connections[pod]; ok {
+		return conn, nil
+	}
+
+	transport, upgrader, err := spdy.RoundTripperFor(p.config)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := p.config
+	cfg.APIPath = "/api"
+	cfg.GroupVersion = &schema.GroupVersion{Version: "v1"}
+	cfg.NegotiatedSerializer = negotiatedSerializer
+	restClient, err := rest.RESTClientFor(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	req := restClient.Post().
+		Resource("pods").
+		Namespace(pod.Namespace).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, req.URL())
+	conn, protocol, err := dialer.Dial(portForwardProtocolV1Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if protocol != portForwardProtocolV1Name {
+		if conn != nil {
+			conn.Close()
+		}
+
+		return nil, fmt.Errorf("unable to negotiate protocol: client supports %q, server returned %q", portForwardProtocolV1Name, protocol)
+	}
+
+	refCountedConn := &refCountedConnection{Connection: conn}
+
+	p.connections[pod] = refCountedConn
+	return refCountedConn, nil
+}
+
+type conn struct {
+	dataStream  httpstream.Stream
+	errorStream httpstream.Stream
+	network     string
+	remote      string
+	onClose     func()
+
+	errCh  chan error
+	stopCh chan error
+	closed atomic.Bool
+}
+
+var _ net.Conn = (*conn)(nil)
+
+func wrapConn(ctx context.Context, onClose func(), network, remote string, s, err httpstream.Stream) *conn {
+	c := &conn{
+		dataStream:  s,
+		errorStream: err,
+		network:     network,
+		remote:      remote,
+		onClose:     onClose,
+		errCh:       make(chan error, 1),
+		stopCh:      make(chan error),
+	}
+
+	go c.pollErrors()
+	go c.checkCancelation(ctx)
+
+	return c
+}
+
+func (c *conn) checkCancelation(ctx context.Context) {
+	select {
+	case <-ctx.Done():
+		c.writeError(ctx.Err())
+		c.Close()
+	case <-c.stopCh:
+	}
+}
+
+func (c *conn) pollErrors() {
+	defer c.Close()
+
+	data, err := io.ReadAll(c.errorStream)
+	if err != nil {
+		c.writeError(err)
+		return
+	}
+
+	if len(data) != 0 {
+		c.writeError(fmt.Errorf("received error message from error stream: %s", string(data)))
+		return
+	}
+}
+
+func (c *conn) writeError(err error) {
+	select {
+	case c.errCh <- err:
+	default:
+	}
+}
+
+func (c *conn) checkError() error {
+	select {
+	case err := <-c.errCh:
+		return err
+	default:
+		if c.closed.Load() {
+			return net.ErrClosed
```

Comment:
What was the reason for adding this check? It seems that omitting it would result in _roughly_ the same behavior, depending on the behavior of `.Read`/`.Write` on a closed stream.